### PR TITLE
run: fix Linux tunnel-backed endpoints in per-process mode

### PR DESCRIPTION
**Pull request by magurotuna**
Addresses the Linux side of #765. The macOS `getaddrinfo` / mDNSResponder bypass is tracked separately in #772.

## Summary

In Linux per-process mode, tunnel-backed endpoints are reachable through DNS-VIP interception. If the wrapped process resolves a hostname through a host-side resolver path instead, it receives the raw upstream address and the gateway cannot dispatch the connection through the configured tunnel.

This PR makes the child namespace's DNS policy fail closed, fixes TCP connection-failure propagation, and makes IPv6 DNS-VIP answers routable.

### Fail-closed DNS lockdown

- Bind-mount the gateway resolver configuration, a minimal synthetic `/etc/hosts`, and—when necessary—a canonicalized `/etc/nsswitch.conf` into the child mount namespace.
- Recognize glibc-valid `hosts` definitions, including whitespace around the colon, colon-less forms, and duplicates. Unsafe definitions and action modifiers are reduced to the deterministic `files dns` chain.
- Remount all lockdown files read-only. Setup aborts if a required bind or read-only remount fails.
- Mask systemd-resolved's varlink socket and existing nscd runtime directories. The nscd directory mask is an empty read-only tmpfs, so a socket recreated while the wrapped process is running remains hidden.
- Treat unexpected discovery errors as fatal rather than silently skipping a mask.
- Preserve `CLAWPATROL_RUN_KEEP_RESOLV=1` as the explicit escape hatch that skips the lockdown.

A narrow residual remains when no nscd runtime directory exists during setup and nscd creates one later; this is documented in the implementation.

### Dial upstream before accepting the child-side TCP connection

The per-session forwarder now performs `transport.Dial` with a 15-second bound before completing the child-side TCP handshake. An upstream dial failure is returned to the wrapped process as a refused connection rather than a successful `connect()` followed by an application-handshake timeout. Clients that implement A/AAAA fallback can then continue to another address.

### Route IPv6 DNS-VIP traffic

The child namespace receives an IPv6 address and a route for the DNS-VIP prefix `fd78::/64`. The route is deliberately scoped rather than an IPv6 default route because the TUN bridge does not yet forward IPv6 UDP; broader IPv6 UDP support is tracked in #781. Hosts booted with `ipv6.disable=1` warn and retain the IPv4 path.

The existing gateway transports already intercept UDP/53, so no separate daemon-side resolver is required.

## Testing

- Planner and NSS unit tests cover Fedora/systemd-resolved, Ubuntu, musl, nscd masking, discovery failures, the escape hatch, glibc-compatible `hosts` grammar, duplicates, and action modifiers.
- Integration tests re-exec into a real user and mount namespace and verify live Unix-socket/directory masking, read-only file mounts, and fail-closed error propagation. They self-skip where unprivileged user namespaces are unavailable.
- A two-stack gVisor harness covers refused connections and bidirectional relay over IPv4 and an `fd78::` IPv6 destination, including race-detector runs.
- A live Ubuntu 24.04 E2E on exact head `a1d800c4` passed the unprivileged namespace path: DNS-VIP A/AAAA, scoped IPv6 routing, read-only resolver mounts, systemd-resolved/nscd bypass defenses, IPv6 TCP through a `local_command` PostgreSQL endpoint, and direct passthrough. See the [verification comment](https://github.com/denoland/clawpatrol/pull/771#issuecomment-5012756332).

### CODE REVIEW

**Comment by magurotuna** [comment]
## Adversarial review — changes requested

Reviewed exact head `afc13ddd5802a6242a1e942085ab0709155af393` with three independent Hermes reviewers, Claude Code/Fable, targeted tests, CI-log inspection, and independent source tracing.

### Blocking findings

1. **The DNS lockdown still has an nscd bypass** (`cmd/clawpatrol/run_dns_lockdown_linux.go:141-142`).

   glibc's `getaddrinfo` attempts `__nscd_getai` before traversing the `hosts:` NSS database. On a host with hosts caching enabled in nscd/unscd, `/var/run/nscd/socket` remains visible in the child mount namespace, so the host daemon can return a host-resolved/raw address before the bind-mounted `resolv.conf`, sanitized `nsswitch.conf`, or synthetic `/etc/hosts` is consulted. For a tunnel-backed endpoint this bypasses DNS-VIP allocation and reintroduces the black-hole behavior this PR is intended to fail closed against.

   I verified this control flow against glibc's `getaddrinfo.c`, `nscd_getai.c`, and `nscd_helper.c`. Please detect and mask `/var/run/nscd/socket` (normally `/run/nscd/socket`) using the same empty-file bind-mount mechanism as the systemd-resolved varlink socket, and test it with an actual Unix socket.

2. **Other host-backed NSS sources are still allowed to answer before gateway DNS** (`cmd/clawpatrol/run_linux.go:994-1052`).

   `rewriteHostsLine` removes only `resolve` and `mdns*`, while preserving sources such as `sss`, `ldap`, and `nis` in their original positions. For example, `hosts: files sss dns` is deliberately left unchanged. If an earlier source answers for a tunnel-backed hostname, glibc never reaches the gateway-controlled `dns` source and the child receives the raw address.

   In lockdown mode, the safe deterministic chain should be `files dns`, or at minimum `dns` must precede every resolver source outside the namespace's control. The existing `CLAWPATROL_RUN_KEEP_RESOLV=1` escape hatch can preserve host NSS behavior for users who explicitly require it.

3. **The required lint gate is currently failing on PR-added code** (`cmd/clawpatrol/daemon_session_linux_test.go:35,119,136`).

   Both the live CI log and a local `golangci-lint` run report three `revive` unused-parameter errors. These appear fixable by renaming the unused parameters to `_`.

### Test coverage issue

`TestRunStackTCPForwarderDialFailureRefusesConnect` does not distinguish a fast RST from a stranded SYN that reaches its 10-second context deadline. It only checks that `err != nil`.

I verified this by temporarily removing `req.Complete(true)` in an isolated verification worktree: the test still passed after exactly 10 seconds. Please assert the connection-refused error (and preferably a comfortably sub-deadline completion time). The new forwarder tests should also be parameterized over IPv4 and an `fd78::` destination so the new IPv6 path is exercised.

### Verification performed

- Targeted Go tests passed at the reviewed head.
- `make dashboard` passed.
- Local lint reproduced the same three CI errors.
- Review worktree remained unchanged, including ignored/generated state.
- Live PR head was rechecked immediately before posting this comment.

Given the two remaining resolver-bypass paths, I do not think the current implementation satisfies its stated fail-closed invariant yet.

**Comment by magurotuna** [comment]
All review findings addressed on branch [`claude/pr771-review-comments-1uly39`](https://github.com/denoland/clawpatrol/tree/claude/pr771-review-comments-1uly39) (commit b980deb, based on this PR's head and ready to merge into `fix/765-per-process-dns-lockdown`).

### Blocking findings

1. **IPv6 default route black-holing UDP/QUIC** — the child netns no longer gets a v6 default route. The route is scoped to the DNS-VIP prefix `fd78::/64` (`dnsvip.DefaultCIDR6`), which is the only v6 traffic the bridge needs to carry (VIP endpoints are TCP by construction — DNS rides the v4 nameserver). Every other AAAA destination, TCP and UDP alike, now fails with instant network-unreachable so clients fall back to IPv4 rather than stalling on silently dropped IPv6 UDP.

2. **nscd bypass** — the lockdown planner now masks `/run/nscd/socket` and `/var/run/nscd/socket` (both spellings; on most distros they're the same inode via the `/var/run` symlink, and the stacked bind-mount is harmless) with the same empty-file bind-mount used for the resolved varlink socket. Since `__nscd_getai` runs before the NSS chain is traversed, masking the socket is the only layer that can close this. Planner unit tests cover the new matrix entries, and the namespace integration test now creates a **live unix socket with an accepting listener**, masks it, and asserts that `connect()` fails through the mask while the listener is still alive. (The canonical `/run/nscd/socket` path can't be created by an unprivileged uid-mapped test, so the mechanism is exercised at a temp path and the canonical paths are pinned by the planner tests.)

3. **Non-fail-closed NSS chain** — `rewriteHostsLine` is now an allowlist reduced to `files` + `dns` (original order preserved, `dns` appended when absent). `sss`, `ldap`, `nis`, `mymachines`, `myhostname`, and anything else off the allowlist is dropped along with trailing action brackets. `myhostname` is intentionally dropped too: self-lookups are served by the synthetic `/etc/hosts` through `files`. Hosts that resolve internal names via sssd/LDAP lose that inside the sandbox; `CLAWPATROL_RUN_KEEP_RESOLV=1` remains the documented escape hatch. The tests that previously locked in the unsafe ordering now assert the opposite.

4. **Lint** — the three revive unused-parameter errors in `daemon_session_linux_test.go` are fixed; `golangci-lint 2.12.2` (CI's version) reports 0 issues on the package.

5. **`Fixes #765` scope** — narrowed the closing reference in the PR description; #765 stays open to track the macOS `getaddrinfo`/mDNSResponder bypass, which this PR doesn't touch.

### Test-coverage finding

`TestRunStackTCPForwarderDialFailureRefusesConnect` now asserts a connection-refused error **and** completion well under the 10s dial deadline, and both forwarder tests are parameterized over IPv4 and an `fd78::` destination. Verified by mutation: removing `req.Complete(true)` makes both subtests fail with `connect error = context deadline exceeded; want connection refused` after 10s.

### Verification

- `gofmt` clean; `golangci-lint run ./cmd/clawpatrol/...` → 0 issues.
- Targeted lockdown/netns/forwarder tests pass; forwarder tests clean under `-race -count=3`.
- Pre-existing, unrelated: `TestStartWGServerHonorsListenPort` fails in my sandbox identically with and without these changes (UDP bind restriction in the environment).
- Still outstanding from the PR description: end-to-end verification on a real Linux client against a gateway with a tunnel-backed endpoint.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cmd/clawpatrol/run_linux.go [hunk 1]
```diff
@@ -887,16 +891,49 @@ func recvFDs(s *os.File, n int) ([]int, error) {
...
+// v6 bring-up degrades AAAA attempts to instant network-unreachable,
+// which clients fall back from.
+func childNetnsSteps(tunAddr string) []netnsStep {
+	return []netnsStep{
+		{args: []string{"ip", "link", "set", "lo", "up"}},
+		{args: []string{"ip", "link", "set", tunIfName, "mtu", fmt.Sprintf("%d", runTunMTU), "up"}},
+		{args: []string{"ip", "addr", "add", tunAddr + "/32", "dev", tunIfName}},
+		{args: []string{"ip", "route", "add", "default", "dev", tunIfName}},
+		{args: []string{"ip", "-6", "addr", "add", runTunAddr6 + "/128", "dev", tunIfName, "nodad"}, optional: true},
+		{args: []string{"ip", "-6", "route", "add", "default", "dev", tunIfName}, optional: true},
```
[the remark this answers is not in the document]
Comment (magurotuna) [reply cmd/clawpatrol/run_linux.go]: Fixed in c9830f2 (merge of `claude/pr771-review-comments-1uly39`): the v6 default route is gone — the child now gets a route scoped to the DNS-VIP prefix only (`ip -6 route add fd78::/64 dev wg0`, from `dnsvip.DefaultCIDR6`). VIP endpoints are TCP by construction and DNS rides the v4 nameserver, so that prefix is the only v6 traffic the bridge needs to carry; every other AAAA destination (TCP and UDP alike) now fails with instant network-unreachable and clients fall back to IPv4 instead of stalling on silently dropped v6 UDP. `TestChildNetnsSteps` pins the scoped route, and the forwarder tests are parameterized over v4 and an fd78:: destination.

File: cmd/clawpatrol/run_dns_lockdown_linux.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		})
+	}
+	if in.HostsExists {
+		plan.Overrides = append(plan.Overrides, etcOverride{
+			Target:  "/etc/hosts",
+			Pattern: "clawpatrol-hosts-*",
+			Body:    minimalHostsFile(in.Hostname),
+		})
+	}
+	if in.VarlinkSocketExists {
```
[the remark this answers is not in the document]
Comment (magurotuna) [reply cmd/clawpatrol/run_dns_lockdown_linux.go:181]: Fixed in c9830f2: the planner now masks `/run/nscd/socket` and `/var/run/nscd/socket` (both spellings — usually the same inode via the `/var/run` symlink; the stacked bind is harmless) with the same empty-file mechanism as the resolved varlink socket. Since `__nscd_getai` runs before the NSS chain is traversed, socket masking is indeed the only layer that can close this. The namespace integration test now creates a live unix socket with an accepting listener, masks it, and asserts `connect()` fails through the mask while the listener stays alive; the canonical paths (which an unprivileged uid-mapped test can't create) are pinned by planner unit tests.

File: cmd/clawpatrol/run_dns_lockdown_linux.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	plan.Overrides = append(plan.Overrides, etcOverride{
+		Target:  "/etc/resolv.conf",
+		Pattern: "clawpatrol-resolv-*",
+		Body:    in.ResolvBody,
+	})
+	if in.NsswitchErr != nil {
+		if !errors.Is(in.NsswitchErr, fs.ErrNotExist) {
+			return dnsLockdown{}, fmt.Errorf("read /etc/nsswitch.conf: %w", in.NsswitchErr)
+		}
+	} else if body, changed := rewriteHostsLine(in.NsswitchRaw); changed {
```
[the remark this answers is not in the document]
Comment (magurotuna) [reply cmd/clawpatrol/run_dns_lockdown_linux.go]: Fixed in c9830f2: `rewriteHostsLine` is now an allowlist reduced to `files` + `dns` (original order preserved, `dns` appended when absent) — `sss`, `ldap`, `nis`, `mymachines`, `myhostname` and anything else off the list is dropped along with trailing action brackets. `myhostname` goes too because self-lookups are served by the synthetic `/etc/hosts` through `files`. Hosts resolving internal names via sssd/LDAP lose that inside the sandbox; `CLAWPATROL_RUN_KEEP_RESOLV=1` stays as the documented escape hatch. The tests that previously locked in the old ordering now assert the opposite.

**Comment by magurotuna** [comment]
All review findings are now on this branch: merged `claude/pr771-review-comments-1uly39` as c9830f2 (the one conflict — `daemon_session_linux_test.go` between the earlier lint fix 6f6f08f and the reworked forwarder tests — was resolved in favor of the reworked tests, which include the lint renames).

Status of the two general points from the review:

- **Lint**: fixed (6f6f08f, superseded by the reworked tests). `golangci-lint v2.12.2` (CI's version) reports 0 issues on the package.
- **`Fixes #765` scope**: the closing reference is narrowed to "Addresses the Linux side of #765". The macOS `getaddrinfo`/mDNSResponder bypass also has a dedicated follow-up, #772 — so after this merges we can either keep #765 open for the macOS half or close it in favor of #772, whichever tracking style is preferred.

Verification after the merge: build + `gofmt` clean; all lockdown/netns/forwarder tests pass in a Linux container, including the namespace integration tests (live-socket mask) under `--privileged`; forwarder tests clean under `-race -count=3`.

Still outstanding (as noted in the description): end-to-end verification on a real Linux client against a gateway with a tunnel-backed endpoint.

**Comment by magurotuna** [comment]
## Follow-up adversarial review of `c9830f2c`

The previously reported findings are addressed at this head: the canonical NSS chain is reduced to `files`/`dns`, nscd sockets present during setup are masked, lint is green, the refusal test now rejects deadline expiry, IPv6 is covered by the forwarder harness, and the integration test masks a live Unix socket.

I found two remaining fail-closed gaps.

### 1. `rewriteHostsLine` accepts a narrower grammar than glibc

`cmd/clawpatrol/run_linux.go:1040-1080`

The rewriter recognizes only an exact `hosts:` prefix and returns after the first matching line. glibc's `nss/nss_database.c::process_line`, however:

- accepts whitespace before/around the colon (`hosts : ...`);
- accepts a colon-less form (`hosts ...`); and
- processes duplicate database definitions by assigning each parsed result, so the last `hosts` line wins.

For example:

```text
hosts: files dns
hosts: sss dns
```

`rewriteHostsLine` sees the first line, decides that no change is needed, and returns. glibc then uses the second line, leaving `sss` ahead of gateway DNS and restoring the raw-address/black-hole path. The same bypass is possible with `hosts : sss dns` or `hosts sss dns`, which the current `CutPrefix(..., "hosts:")` does not recognize at all.

I independently verified these parsing semantics against current glibc `nss/nss_database.c:198-227` (`process_line` accepts spaces/colons and unconditionally assigns `data->services[db]`). All three follow-up reviewer lanes agreed on the implementation/test gap.

Please mirror glibc's database-name grammar and sanitize every effective `hosts` definition (or replace all such definitions with one deterministic `hosts: files dns` line). Add explicit tests for duplicate lines, space-before-colon, and colon-less forms; the test helper should validate the complete resulting set of `hosts` entries rather than only the first one.

### 2. nscd discovery still has fail-open cases

`cmd/clawpatrol/run_dns_lockdown_linux.go:175-180`

The discovery loop treats every `os.Stat` error as "socket absent":

```go
if _, err := os.Stat(p); err == nil {
    nscd = append(nscd, p)
}
```

A permission/I/O error therefore silently skips the mask instead of aborting. Also, when nscd is absent during discovery, no mount is established; if nscd/unscd creates its socket later, the child mount namespace can see the new socket and glibc can again consult it before the sanitized NSS chain.

At minimum, propagate all errors other than `ENOENT` so unexpected discovery failures are fail-closed. For the create-after-discovery case, please either establish a persistent mask at the nscd runtime-directory level where possible, or explicitly redesign/document this as a residual limitation rather than claiming complete nscd fail-closed coverage. Add a focused test for non-`ENOENT` discovery failure and, if supported by the chosen design, socket creation after lockdown application.

### Verification status

- Exact reviewed head: `c9830f2cf9485b0825e240b47d96b0d76b1d3f9a`
- GitHub lint/test/sandbox checks: all green
- Focused Go tests under `-race -count=3`: passed
- Local lint rerun was interrupted by VM disk exhaustion during dependency typechecking; the corresponding GitHub lint job passed at this exact head
- Review worktree remained read-only

Given finding 1, the core allowlist can still be bypassed by glibc-valid host configuration, so I recommend keeping changes requested.

**Comment by magurotuna** [comment]
## Follow-up adversarial review — two remaining blockers

I re-reviewed current head `c9830f2cf9485b0825e240b47d96b0d76b1d3f9a`. The previously reported findings (nscd masking, non-local NSS providers, IPv6 route scope, refusal assertions/IPv6 coverage, lint, and issue scope) are now addressed. I found two remaining fail-closed gaps:

### 1. The wrapped command can overwrite the bind-mounted lockdown files

`bindOverEtc` creates each source inode with `os.CreateTemp`, changes it to `0644`, and bind-mounts it without a read-only remount (`cmd/clawpatrol/run_linux.go:1085-1116`). On the unprivileged-user-namespace path, that inode is owned by the mapped invoking UID, and the final wrapped command runs as the same UID. Clearing ambient capabilities does not remove normal owner write permission.

I reproduced this in a fresh `CLONE_NEWUSER | CLONE_NEWNS` helper matching the integration-test setup: after `bindOverEtc`, an ordinary `os.WriteFile(target, ...)` from the mapped UID successfully replaced the mounted content. This means a wrapped process can rewrite `/etc/hosts`, `/etc/resolv.conf`, or the sanitized `/etc/nsswitch.conf` after setup and reintroduce a raw-IP resolver path. The sudo path is not affected in the same way because its temporary inode is root-owned, but the default unprivileged path is.

Please make these bind mounts immutable to the wrapped process, e.g. perform the bind and then remount with `MS_REMOUNT | MS_BIND | MS_RDONLY` (failing closed if that step fails), and add an integration assertion that writes and chmod attempts fail after setup.

### 2. Preserved NSS action modifiers can still prevent gateway DNS from running

The new source allowlist is directionally correct, but `rewriteHostsLine` still preserves bracketed action modifiers attached to surviving `files`/`dns` entries (`cmd/clawpatrol/run_linux.go:1052-1058`). It also returns `changed=false` when the source names are already allowlisted, even if their actions are unsafe.

For example:

```text
hosts: files [NOTFOUND=return] dns
```

is treated as an unchanged safe chain. For any name absent from the synthetic `/etc/hosts`, `nss_files` returns `NOTFOUND`, the action immediately returns from the hosts database, and the gateway-backed `dns` source is never consulted. That violates the stated invariant that tunnel-backed lookups must reach gateway DNS.

I confirmed the current function leaves this exact input untouched. In lockdown mode, please canonicalize the line to a known-safe chain such as exactly `hosts: files dns`, or rebuild only explicitly safe actions rather than preserving host-provided modifiers.

### Verification

Both focused reproductions pass against the current code, demonstrating the unsafe behavior:

- owner write through the bind-mounted lockdown target succeeds in a user+mount namespace;
- `rewriteHostsLine("hosts: files [NOTFOUND=return] dns\n")` reports no change.

Given that either path can bypass or suppress the gateway resolver, I believe both should remain blocking for the PR's fail-closed claim.

**Comment by magurotuna** [comment]
All four findings from the follow-up reviews are addressed on branch [`claude/pr771-review-comments-1uly39`](https://github.com/denoland/clawpatrol/tree/claude/pr771-review-comments-1uly39) (commit a291bee, based on this PR's head `c9830f2` — fast-forward mergeable into `fix/765-per-process-dns-lockdown`).

### 1. nsswitch parsing now mirrors glibc's grammar, all definitions sanitized

`rewriteHostsLine` is replaced by `cutHostsDefinition` + `rewriteHostsLines`. The recognizer follows `nss/nss_database.c::process_line`: leading whitespace, name terminated by whitespace **or** colon, separator being any run of whitespace/colons — so `hosts : sss dns`, `hosts:sss dns`, and the colon-less `hosts sss dns` are all recognized (a bare `hosts` is a glibc syntax error and stays ignored, matching glibc). Every definition is processed, not just the first, so with duplicates the last-wins line glibc actually honors is always sanitized. Unsafe definitions are canonicalized to exactly `hosts: files dns`. New tests cover duplicate lines (safe first + unsafe second), space-before-colon, colon-less (both safe and unsafe), bare `hosts`, and empty `hosts:`, and the test helper now validates the **complete** set of hosts definitions rather than the first.

### 2. Action modifiers no longer survive

A definition counts as safe only when every token is an allowlisted source (`files`/`dns`), `dns` is present, and there are **no** bracketed modifiers. Your exact repro `hosts: files [NOTFOUND=return] dns` is now rewritten to `hosts: files dns` (test: `action modifier on kept source dropped`). Modifiers are dropped wholesale rather than audited — the default actions are exactly what the files→dns chain needs.

### 3. Lockdown mounts are now immutable to the wrapped command

`bindOverEtc` finishes with `remountReadOnly`: `MS_REMOUNT|MS_BIND|MS_RDONLY`, re-asserting the mount's existing `nosuid`/`nodev`/… flags via statfs (required in a userns, where a remount that drops inherited flags gets EPERM — the temp inode typically lives on a nosuid,nodev /tmp). Failure of the remount is fatal (fail closed). The namespace integration test asserts that writes **and** chmod fail on `/etc/resolv.conf`, `/etc/hosts`, and both mask types — and I verified by mutation that removing the remount makes the test fail with `write to /etc/resolv.conf succeeded`. Unmounting from inside is not available to the agent either: ambient caps are cleared before exec, and mounts inherited into any namespace it creates itself are kernel-locked (MNT_LOCK).

### 4. nscd discovery is fail-closed; directory-level mask

- Discovery now aborts on any stat failure other than ENOENT/ENOTDIR (`classifyStatErr`, unit-tested including the EPERM/unknown-error fatal paths). `gatherDNSLockdownInputs` returns an error and `runRunChild` treats it like any other lockdown failure. The `CLAWPATROL_RUN_KEEP_RESOLV=1` escape hatch short-circuits **before** discovery so it still works on hosts where the probes themselves fail.
- The mask moved from the socket file to the runtime directory: an existing `/run/nscd` (and `/var/run/nscd`) is hidden behind an empty **read-only tmpfs**, so a socket re-created mid-run by a restarting daemon lands in the host directory and stays invisible to the child; the read-only tmpfs also prevents the same-uid agent from planting a fake socket. The integration test dir-masks a live listening unix socket and asserts it disappears and the shadow is unwritable.
- Residual, now documented in the file-top comment: if **no** nscd directory exists at setup (nscd installed but stopped) and the daemon starts mid-run, it creates `/run/nscd` in the shared `/run` and the child will see it — there is no path to mount over up front and the unprivileged child cannot create one in the host-owned `/run`. The directory mask closes the mid-run window only when nscd is already running at setup.

### Verification

- `gofmt` clean; `golangci-lint 2.12.2` → 0 issues.
- All lockdown/nsswitch/netns/forwarder tests pass, including the namespace integration tests (which exercise the real remount + tmpfs behavior in an unprivileged user+mount namespace).
- Pre-existing, unrelated: `TestStartWGServerHonorsListenPort` still fails in my sandbox with and without these changes (UDP bind restriction).
- Still outstanding: end-to-end verification on a real Linux client against a gateway with a tunnel-backed endpoint.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by magurotuna** [comment]
The second-round findings are now on this branch: merged `claude/pr771-review-comments-1uly39` (a291bee) as a1d800c — clean merge on top of the earlier main sync (d8198ef), #766's CA-var changes intact.

Per finding:

1. **glibc nsswitch grammar** — `cutHostsDefinition` now mirrors `nss/nss_database.c::process_line` (whitespace around the colon, colon-less form), and `rewriteHostsLines` canonicalizes **every** `hosts` definition to `hosts: files dns` (last-one-wins duplicates included), leaving a line untouched only when it is already allowlist-only with `dns` present and no action modifiers.
2. **Action modifiers on allowlisted sources** — host-provided brackets no longer survive: `hosts: files [NOTFOUND=return] dns` is now rewritten (the default NOTFOUND=continue is exactly what the files→dns chain needs).
3. **Writable bind mounts** — `bindOverEtc` now finishes with `MS_REMOUNT|MS_BIND|MS_RDONLY` (re-asserting the mount's existing nosuid/nodev/… flags, which the kernel requires for a userns remount), failing closed if the flip fails. The integration test asserts writes fail after setup.
4. **nscd discovery fail-open** — `gatherDNSLockdownInputs` now returns an error on unexpected stat failures instead of silently dropping a mask; `CLAWPATROL_RUN_KEEP_RESOLV=1` short-circuits before any probe so the escape hatch works even when discovery itself fails.

Verification after the merge: build + `gofmt` clean; `golangci-lint v2.12.2` → 0 issues; all lockdown/netns/forwarder tests pass in a Linux container including the namespace integration tests under `--privileged`; the only package failures are the three pre-existing container-environment ones (`TestExampleSocksTunnel`, `TestPluginTunnelVia*`) that fail identically on `main`.

**Comment by magurotuna** [comment]
## Final re-review — ready to land

Re-reviewed current head `a1d800c4b6ca595493d87478a9c5a8f81ad6cbad` against the previously reported findings.

### Findings resolved

- Lockdown file bind mounts are now remounted read-only, with fail-closed error propagation. The namespace integration test verifies that both writes and chmod attempts fail.
- NSS handling now recognizes glibc's accepted `hosts` syntaxes, rewrites every duplicate definition, removes all action modifiers, and canonicalizes unsafe definitions to `hosts: files dns`.
- nscd is masked at the runtime-directory level with an empty read-only tmpfs, covering socket recreation while the directory exists at setup.
- Unexpected discovery/stat errors now fail closed.
- The earlier nscd, non-local NSS provider, IPv6 route, TCP refusal assertion, IPv6 test coverage, lint, and issue-scope findings remain addressed.

### Verification

- All GitHub checks are green: lint, test, Linux Landlock sandbox, macOS sandbox, and preview.
- Focused DNS planner/NSS/forwarder tests passed locally under the race detector.
- `git diff --check origin/main...HEAD` is clean.
- The namespace integration test does not run successfully inside this review environment because the environment is itself nested behind an unlinked `/etc/resolv.conf` bind mount; the same committed integration test passes in CI's normal Linux environment and exercises the read-only file masks and directory mask.

The documented edge case where no nscd runtime directory exists at setup and a daemon creates it later remains a narrow accepted limitation rather than a regression introduced by this fix.

**Verdict: no remaining blocking findings; ready to land.**

Note: the PR is still shown as blocked because the earlier `CHANGES_REQUESTED` review remains active. This comment is posted through the PR author's GitHub identity, which cannot approve its own PR; the previous review must be dismissed or updated by an eligible reviewer/maintainer.

**Comment by magurotuna** [comment]
Ubuntu live E2E on exact head `a1d800c4b6ca595493d87478a9c5a8f81ad6cbad`: **PASS** for the supported Linux tunnel-backed TCP path.

- Disposable Ubuntu 24.04.4 QEMU/KVM VM (kernel 6.8, glibc 2.39), non-root `uid=1000`.
- Forced the normal unprivileged user/mount/network-namespace path with `CLAWPATROL_NO_SUDO=1`; the guest-only AppArmor userns restriction was disabled so that path could run.
- Static Linux/amd64 binary SHA-256: `4b7505cfa9a379ca63f39b3e35872993dc8665df97fd6261c0f7d5e4652753e2`.
- Fresh per-process tsnet identity using profile `yusuke` and the existing staging PostgreSQL endpoint backed by `local_command` / Cloud SQL Proxy.

Observed inside `clawpatrol run`:

- `getent` returned `10.78.0.4` and `fd78::4`.
- The child netns had `fd78::/64 dev wg0` and no IPv6 default route through `wg0`.
- Resolver policy was `nameserver <gateway>` plus canonical `hosts: files dns`.
- Writes to `/etc/resolv.conf`, `/etc/hosts`, and `/etc/nsswitch.conf`, and `chmod /etc/resolv.conf`, failed on the read-only mounts.
- The main systemd-resolved varlink socket was replaced and could not be connected to.
- With a real Ubuntu `nscd` active, its host cache was deliberately poisoned with `203.0.113.9`; unwrapped `getent` returned the poisoned entry, while wrapped `getent` ignored it, returned the DNS VIPs, and could not connect to the nscd socket.
- `strace` captured an AF_INET6 connection to `fd78::4:5432`, and `psql` completed `SELECT 1` through the tunnel-backed endpoint.
- Direct passthrough remained healthy (`https://pypi.org/simple/` → HTTP 200, certificate verification successful).
- Killing the per-user daemon and rerunning the unprivileged path still produced the VIP and completed another query.

Limitations: the tailnet-only join emitted the known peer-claim warning tracked by #778 but recovered through daemon registration. No intentionally broken live tunnel endpoint was available, so the dial-before-accept failure path remains covered by the focused tests rather than this live run. The VM host was also using Claw Patrol, so the unwrapped baseline was not a clean non-Claw-Patrol control: it already resolved the test hostname to a VIP. The wrapped-path assertions and successful tunnel-backed query were validated independently.

The VM, disk, SSH key, binary, and local client state were removed after the run. No new blocking finding was observed.
